### ISSUE TITLE
[ls] The incremental builder occasionally deletes generated files

### ISSUE THREAD

**Issue by kittaakos** [eclipse/xtext-core#799]
It does not happen with the clustering builder state in Eclipse.

https://github.com/eclipse/xtext-core/blob/efa0ddbde33a67fb9d685ffc1cb5359d35bc6982/org.eclipse.xtext/src/org/eclipse/xtext/build/IncrementalBuilder.xtend#L175-L179

For the language I am currently working, there is some black magic that out of nothing generates a file inside the gen folder. Now, this is cleaned up by the incremental builder.

I try to come up with a stripped down example (if no, just mention me) soon. I wanted to open this issue before I forget it.

**Comment by cdietrich** [comment]
can you give more details on what the problem is. that files are created and deleted by the builder happens in eclipse too

**Comment by kittaakos** [comment]
The use-case is somewhat related to #801. Multiple files are created on the fly **before** even hitting the generate phase. Perhaps that is the problem. Can you please point me to the code in the Eclipse builder where we delete **stale** generated files? Thanks!

My concern is that the LS incremental builder is trying to be smarter than it should be.

**Comment by cdietrich** [comment]
see org.eclipse.xtext.builder.BuilderParticipant.cleanDerivedResources(Delta, Set<IFile>, IBuildContext, EclipseResourceFileSystemAccess2, IProgressMonitor)

**Comment by cdietrich** [comment]
@kittaakos is it possible you have a n:1 generation scheme: 1 target file gets generated from multiple source file?

**Comment by kittaakos** [comment]
> @kittaakos is it possible you have a n:1 generation scheme: 1 target file gets generated from multiple source file?

No. That is not the case. A model file might end up in multiple generated files.

**Comment by cdietrich** [comment]
but if the condtions (the may) change then it is correct to delete the stale files

**Comment by cdietrich** [comment]
@kittaakos any feedback on the issue?

**Comment by kittaakos** [comment]
What kind of feedback are you looking for? You need the promised use case, right? (Unfortunately, I won't have time to make it this week.) Or is it still unclear when and how does this happen?

**Comment by cdietrich** [comment]
@kittaakos any news on this?

**Comment by kittaakos** [comment]
Unfortunately, there are no updates. Did you hit the same issue, @cdietrich?

**Comment by cdietrich** [comment]
no. this is why i am asking.

**Comment by kittaakos** [comment]
As I [wrote earlier](
https://github.com/eclipse/xtext-core/issues/799#issuecomment-407427475), if files are created outside of the generator (and generation phase), those files will be removed by the incremental builder. This does not happen with the builder in Eclipse.

**Comment by cdietrich** [comment]
yes but i wonder why these files are in previous. eclipse builder does the same
=> there must be a difference how these files end up in previous
